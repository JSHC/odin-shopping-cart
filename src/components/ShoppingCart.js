import '../styles/ShoppingCart.css';

const ShoppingCart = (props) => {
    const {cartItems, onDeleteItem} = props;
    return (
        <div>
            <h1>Shopping Cart</h1>

            <div className='shopping-cart-item-header'>
                <p>Product</p>
                <p>Quantity</p>
                <p>Price</p>
            
                {cartItems.map(item => {
                    return (
                        <div className="shopping-cart-item" key={item.id}>
                            <p>{item.name}</p>
                            <p>{item.quantity}</p>
                            <p>{item.price}</p>
                            <button 
                            aria-label='delete-shopping-cart-item'
                            onClick={() => {onDeleteItem(item)}}>
                            X
                            </button>
                        </div>
                    )
                })}
                {(cartItems && cartItems.length) > 0 ?
                    <p className='shopping-cart-sum'>Sum: {cartItems.reduce((prev, curr) => prev + (curr.quantity * curr.price), 0)}</p>
                    :
                    <p className='shopping-cart-empty-message'>You haven't added any items to the shopping cart.</p>
                }
                
            </div>
            
        </div>
    );
};

export default ShoppingCart;
