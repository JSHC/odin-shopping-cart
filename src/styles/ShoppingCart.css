.shopping-cart-item {
    display: grid;
    grid-column: span 3;
    grid-template-columns: subgrid;
    border-bottom: 1px solid #ddd;
    padding: 1rem 0;
}

.shopping-cart-item:first-of-type {
    border-top: 1px solid #ddd;
}

.shopping-cart-item-header {
    display: grid;
    grid-template-columns: repeat(4, 1fr);    
}

.shopping-cart-item-header>p {
    font-weight: bold;
}

.shopping-cart-item-header p {
    margin: 0;
}

.shopping-cart-item-header .shopping-cart-sum {
    grid-column: 3;
    margin-top: 20px; 
}