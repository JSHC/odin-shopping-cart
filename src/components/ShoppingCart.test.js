import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ShoppingCart from './ShoppingCart';
import userEvent from "@testing-library/user-event";

describe('ShoppingCart component', () => {
    it('renders', () => {
        const cartItems = [];
        const { container } = render(<ShoppingCart cartItems={cartItems} />);
        expect(container).toMatchSnapshot();
    }) 

    it('renders shopping cart item', () => {
        const cartItems = [
            {id: 1, name: 'Item1', quantity: 1, price: 10}
        ];
        render(<ShoppingCart cartItems={cartItems}/>)
        screen.getByText('Item1');
        screen.getByText('1')
        screen.getByText('10')
    })

    it('renders multiple items', () => {
        const cartItems = [
            {id: 1, name: 'Item1', quantity: 1, price: 10},
            {id: 2, name: 'Item2', quantity: 2, price: 20}
        ]

        render(<ShoppingCart cartItems={cartItems}/>)
        screen.getByText('Item1');
        screen.getByText('1');
        screen.getByText('10');

        screen.getByText('Item2');
        screen.getByText('2');
        screen.getByText('20');
    })

    it ('renders correct sum for single item', () => {
        const cartItems = [
            {id: 1, name: 'Item1', quantity: 3, price: 10},
        ]
        render(<ShoppingCart cartItems={cartItems}/>)
        screen.getByText('Sum: 30');
    })

    it ('renders correct sum for multiple items', () => {
        const cartItems = [
            {id: 1, name: 'Item1', quantity: 3, price: 10},
            {id: 2, name: 'Item2', quantity: 2, price: 20}
        ]
        render(<ShoppingCart cartItems={cartItems}/>)
        screen.getByText('Sum: 70');
    })

    it ('renders message when shopping cart is empty', () => {
        const cartItems = [];
        render(<ShoppingCart cartItems={cartItems} />);
        screen.getByText("You haven't added any items to the shopping cart.")
    })

    it ('renders delete item button', () => {
        const cartItems = [
            {id: 1, name: 'Item1', quantity: 3, price: 10}
        ]
        render(<ShoppingCart cartItems={cartItems}/>)
        screen.getByRole('button', {name: 'delete-shopping-cart-item'});
    })

    it ('can click delete item button', async () => {
        const cartItems = [
            {id: 1, name: 'Item1', quantity: 3, price: 10}
        ]
        const user = userEvent.setup();

        const deleteItemMock = jest.fn();
        render(<ShoppingCart cartItems={cartItems} onDeleteItem={deleteItemMock}/>)
        const deleteItemButton = screen.getByRole('button', {name: 'delete-shopping-cart-item'});
        await user.click(deleteItemButton);

        expect(deleteItemMock).toBeCalled();
    })
})