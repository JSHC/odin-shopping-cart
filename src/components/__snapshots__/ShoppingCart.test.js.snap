// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart component renders 1`] = `
<div>
  <div
    class="shopping-cart"
  >
    <h1>
      Shopping Cart
    </h1>
    <div
      class="shopping-cart-item-header"
    >
      <p>
        Product
      </p>
      <p>
        Quantity
      </p>
      <p>
        Price
      </p>
      <p
        class="shopping-cart-empty-message"
      >
        You haven't added any items to the shopping cart.
      </p>
    </div>
  </div>
</div>
`;
